### DISCUSSION TITLE
Send and get more data from the subscription

### DISCUSSION THREAD

**Discussion by jadayash** [imdhemy/laravel-in-app-purchases#182] [Q&A] [answered]
Again and again, thanks so much for this package you have provided for the community.

As you know, when an in-app purchase is done, we should update the client records in our database, even create an invoice, renew the subscription, or send an email.

How it is possible, to send and receive the client_id from the mobile to the backend with the In-app Events? If this isn't possible, is there any workaround solution you have worked on?

I was thinking, about after the purchase is done, the mobile app sends a post request to the backend with the subscription id and the client id. Do not know if this is an efficient solution?

Moreover, what is a receipt? it is a unique identifier? or what it is?

`use Imdhemy\Purchases\Facades\Subscription;

$receiptResponse = Subscription::appStore()->receiptData($receipt)->verifyRenewable();
`
This $receipt variable from where it came from, or what it represents?

Regards

**Comment by imdhemy** [comment, marked as answer]
Thank you for your nice words. I'm glad you like it.

Regarding your question:

It would be better if the providers were allowed to add some payload to the receipt data, As far as I know, Google used to allow this, but it stopped.

And even more, sometimes both of them (Google and Apple) don't send the initial purchase event. I don't know why, but it happens.

**Proposed solution**

As you mentioned, I think the best solution is to receive the receipt data from the mobile application once the user has purchased the product, and then use the Facades (`Product` or `Subscription`) to verify the receipt data.

You can follow the steps in the [documentation](https://imdhemy.com/laravel-iap-docs/docs/app-store/sell-subscriptions#receipt-validation)

**Note**
I deprecated the `verifyRenewable`. Use the `verify` method instead.

```diff
- $receiptResponse = Subscription::appStore()->receiptData($receipt)->verifyRenewable();

+ $receiptResponse = Subscription::appStore()->receiptData($receipt)->verify();
```

**What is a $receipt**

The `$receipt` is the receipt data that you receive from the mobile application. It's the Base64-encoded receipt data.

**How to identify a subscription**

The `verify` method returns a `ReceiptResponse` object, which contains the original transaction id.  It is the id of the initial purchase. It stays the same for the whole subscription lifecycle, including auto-renewals, cancellations, etc.

```php
$user = $request->user(); // or any other way to get the active user

$transactionId = $receiptInfo->getTransactionId();
$originalTransactionId = $receiptInfo->getOriginalTransactionId();

$productId = $receiptInfo->getProductId(); // The same accessor for (Apple) products and subscriptions

// Check if the user has already purchased the product

// Do whatever to maintain your DB consistency

// Entitle the user to the product or subscription
```

**Reply by jadayash** [reply]
Thanks so much for these detailed answers
But the question is, how can i get the $receipt from the event to then pass it into receiptData()

**Reply by imdhemy** [reply]
You need to add an endpoint to receive the receipt data from the mobile application, not from the event.

Once the user finishes the transaction on the mobile application, it will receive the receipt data and in turn, it can send the data to the specified endpoint.

-- EDIT --

- For the initial purchase, use the endpoint.
- For managing the life cycle, use the events.
